File: SafetyProgram.Core.Models.Serialization/XML/ChemicalTableXml.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Xml.Linq;
using SafetyProgram.Core.IO;

namespace SafetyProgram.Core.Models.Serialization
{
    /// <summary>
    /// Defines an implementation of an IChemicalTable<-->XML (de)serializer.
    /// </summary>
    public sealed class ChemicalTableXml : IStorageConverter<IChemicalTable, XElement>
    {
        /// <summary>
        /// Construct an instance of an IChemcialTable<-->XML (de)serializer.
        /// </summary>
        public ChemicalTableXml()
        { }

        /// <summary>
        /// Serialize the chemical table data in memory to an XML format.
        /// </summary>
        /// <param name="data"></param>
        /// <returns></returns>
        public XElement Store(IChemicalTable data)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Deserialize chemical table data into an object format.
        /// </summary>
        /// <param name="data"></param>
        /// <returns></returns>
        public IChemicalTable Load(XElement data)
        {
            string loadedHeader;
            ObservableCollection<ICoshhChemical> loadedChemicals;

            // Header (optional)
            {
                var headerElement = data.Element("header");

                loadedHeader =
                    headerElement != null ?
                        headerElement.Value :
                        "Default Header";
            }

            // Chemicals (optional, table may be empty).
            {
                var chemicalSerializer = new CoshhChemicalXml();

                loadedChemicals = new ObservableCollection<ICoshhChemical>(
                    from chemicalElement in data.Elements("coshhchemical")
                    select chemicalSerializer.Load(chemicalElement));
            }

            return new ChemicalTable(loadedHeader, loadedChemicals);
        }
    }
}

File: SafetyProgram.Core.Models.FSerialization/ChemicalTableXml.fs
module SafetyProgram.Core.Models.FSerialization.ChemicalTableXml

open Helpers
open CoshhChemicalXml
open ConverterInterface
open SafetyProgram.Core.Models
open System.Xml.Linq
open System.Xml
open System

type ChemicalTableXml() = 

    interface IConverter<ChemicalTable, XElement> with

        // Serialize a coshh chemical into an XML format.
        member this.ConvertTo(data : ChemicalTable) = 
            raise (new NotImplementedException())

        // Deserialize a coshh chemical stored in an XML format.
        member this.ConvertFrom(data : XElement) = 
            let header = 
                match data.Element(xname "header") with
                | null -> Some <| "Default Header"
                | headerElement -> Some <| headerElement.Value

            let chemicals = 
                let chemicalConverter = new CoshhChemicalXml() :> IConverter<CoshhChemical, XElement>
                data.Elements(xname "coshhchemical")
                |> Seq.map(chemicalConverter.ConvertFrom)
                |> fun coshhChemicals ->
                    match coshhChemicals |> Seq.exists(fun coshhChemical -> coshhChemical = None) with
                    | true -> None
                    | false -> coshhChemicals |> Seq.map(fun coshhChemical -> coshhChemical.Value) |> Some

            if (header <> None) || (chemicals <> None) then
                Some <| ChemicalTable { Header=header.Value; Chemicals=chemicals.Value; }
            else
                None
